sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala

```diff
@@ -320,24 +320,31 @@ trait DivModLike extends BinaryArithmetic {
 
   override def nullable: Boolean = true
 
-  final override def eval(input: InternalRow): Any = {
-    val input2 = right.eval(input)
-    if (input2 == null || input2 == 0) {
+  private lazy val isZero: Any => Boolean = right.dataType match {
+    case _: DecimalType => x => x.asInstanceOf[Decimal].isZero
+    case _ => x => x == 0
+  }
+
+  private def divByZero: Any = {
+    if (checkOverflow) {
```

Comment:
After some more thoughts, I think we don't need to control overflow behavior and divide 0 behavior separatedly. Let's just rename the flag to `nullOnError`

Reply:
I think `nullOnError`  is inaccurate for overflow behavior of non-ansi mode, cause it just returns corrupted result rather than null. Maybe a separated flag for `DivModLike` and `Pmod` is better.
Considering that there is also a `DivideInterval.checkOverflow`, shall we open a new PR to fix the flag issue?

Reply:
You are right. How about we flip it as `failOnError`?

Reply:
done